### PR TITLE
Split get context data functions

### PR DESCRIPTION
**Pull request by vanlummelhuizen**
I am slowly compacting code in the get_context_data function.

Work in progress.

### CODE REVIEW

**Comment by susanodd** [comment]
@vanlummelhuizen if you think of any way to refine how the dataset languages are involved in the data structures, related to  #1076, for some datasets there is only one language.
I want to modify templates to not show the language name if only one is there. (So working backwards, probably something goes on in the context, too.)
I like the way the 2-char language code is suffixed onto the id's of elements.
We already had it so that there is a language code needed also for ASL (We had to make a settings when we added model translations). But I suspect the code is far behind.
I saw in the refractors above a lot involve datasets and languages.

**Comment by vanlummelhuizen** [comment]
@susanodd There seems to be quite some redundancy in there. I think some of the redundancy also flows to templates.

I am working on it and putting it all in this draft for you and other to review. More on this later.

**Comment by susanodd** [comment]
Not related but just recalled this about code volume.  I tried to use a separate template (via another file) for modals, to reduce code. But Django can't do "trans" things inside those. (It complains about the template).
There are quite a few chunks of template code that iterate over the (selected) dataset languages, for each of the language-based fields in the text search area. (ALL of the search pages have this.)

**Comment by susanodd** [comment]
@vanlummelhuizen I'm working on code to streamline the processing of the "get" form fields.

I'm working on the Reset of Morpheme Search. But lots of other stuff was messed up so I'm doing #1071 for this code.

I wrote a helper function (in query_parameters.py) to obtain the get parameters that are set for the form in order to give this information back to the context.

(There is a lot of code that does that sideways for Gloss Search as you will have noticed.)

I'm going to also process the query this way to cut that MorphemeListView code down as well.
So basically doing similar things, but now focussing on the GET data.

**Comment by vanlummelhuizen** [comment]
As of https://github.com/Signbank/Global-signbank/pull/1078/commits/76e2fe8f9bd84417fbb746c0baba53b53c8d40f0 the function `get_context_data_for_list_view` fits in my PyCharm screen. :partying_face: 

Next up: 
* rearranging code and removing empty line if appropriate
* searching for redundancy in what data is put in the context

**Comment by vanlummelhuizen** [comment]
This draft's sole purpose is to reduce number of lines per function, and with that, making the code (hopefully) more maintainable. There are lots of commits, so if you are interested in what things I did to reduce the number of lines per function, you best go over this PR one commit at a time. 

Do you agree with what I did? Is it more readable and maintainable?

**Comment by susanodd** [comment]
Nice, @vanlummelhuizen !

**Comment by Woseseltops** [comment]
Wow, I see lot fewer rules and simultaneously more docstrings, impressive. Assuming this does the same thing, this is a huge improvement, in particular because the resulting code looks elegant to me eyes. I tried to find places where it could be improved even furter, and could only find 2:

As Susan knows, I like to keep the happy path unindented, so instead of 
```python
def get_morpheme_idgloss(query_parameters):
    if 'morpheme' in query_parameters.keys():
        try:
            return Morpheme.objects.get(pk=query_parameters['morpheme']).idgloss
        except ObjectDoesNotExist:
            return ''
    return ''
```

it could be
```python
def get_morpheme_idgloss(query_parameters):
    if 'morpheme' not in query_parameters.keys():
        return ''

    try:
        return Morpheme.objects.get(pk=query_parameters['morpheme']).idgloss
    except ObjectDoesNotExist:
        return ''

```

And perhaps this could be drier:

```python
    context['input_names_fields_labels_handedness'] = [
        ('weakdrop', search_form['weakdrop'], search_form['weakdrop'].label),
        ('weakprop', search_form['weakprop'], search_form['weakprop'].label)
    ]
    context['input_names_fields_labels_domhndsh'] = [
        ('domhndsh_letter', search_form['domhndsh_letter'], search_form['domhndsh_letter'].label),
        ('domhndsh_number', search_form['domhndsh_number'], search_form['domhndsh_number'].label)
    ]
    context['input_names_fields_labels_subhndsh'] = [
        ('subhndsh_letter', search_form['subhndsh_letter'], search_form['subhndsh_letter'].label),
        ('subhndsh_number', search_form['subhndsh_number'], search_form['subhndsh_number'].label)
    ]
```

something like

```python

def get_input_names_field_labels(fields):
   return [(field_a, search_form[field_a], search_form[field_a].label) for field in fields]

context['input_names_fields_labels_handedness'] = get_input_names_field_labels(['weakdrop','weak_prop'])
context['input_names_fields_labels_domhndsh'] = get_input_names_field_labels(['domhndsh_letter','domhndsh_number'])
context['input_names_fields_labels_subhndsh'] = get_input_names_field_labels(['subhndsh_letter','subhndsh_number'])
```

**Comment by susanodd** [comment]
Nice, @Woseseltops 


```
def get_input_names_field_labels(fields):
   return [(field_a, search_form[field_a], search_form[field_a].label) for field in fields]

context['input_names_fields_labels_handedness'] = get_input_names_field_labels(['weakdrop','weak_prop'])
context['input_names_fields_labels_domhndsh'] = get_input_names_field_labels(['domhndsh_letter','domhndsh_number'])
context['input_names_fields_labels_subhndsh'] = get_input_names_field_labels(['subhndsh_letter','subhndsh_number'])
```

**Comment by susanodd** [comment]
@vanlummelhuizen I'm happy if you merge this after making the changes of @Woseseltops.
I want to use the code already in the streamline branch (other pull request). The MorphemeListView also has problems with the context variables. Except I have reworked the "search_form" variable.

**Comment by vanlummelhuizen** [comment]
> I tried to find places where it could be improved even furter, and could only find 2:

For learning and fun, discussion and ... inspiration! Good ideas, @Woseseltops ! I will use them.

**Comment by vanlummelhuizen** [comment]
@Woseseltops The second part of your proposal turned out a bit less nice for the eyes because of a missing argument. See https://github.com/Signbank/Global-signbank/pull/1078/commits/05959114f289d23fc742652542ca63cf727c9852

@susanodd If you agree with the latest two commits, feel free to merge.

**Comment by susanodd** [comment]
@vanlummelhuizen question:

https://github.com/Signbank/Global-signbank/blob/05959114f289d23fc742652542ca63cf727c9852/signbank/dictionary/context_data.py#L209

should that be at the end of the file?

**Comment by vanlummelhuizen** [comment]
> should that be at the end of the file?

I don't think it technically matters where it is in the file. Do you think it is more readable/maintainable if it is elsewhere?

**Comment by susanodd** [comment]
No, it's okay for me. @Woseseltops had it as an inline function inside the other function.
